### PR TITLE
Enabled "read with streaming" unit test for browsers that support streaming

### LINKED ISSUE #8851: Enable "read with streaming" test in test/unit/network_spec.js
The "read with streaming" test looks like this:
https://github.com/mozilla/pdf.js/blob/a4cc85fc5f5cb144b6ee7a7978dc83e906677233/test/unit/network_spec.js#L67-L77

There are two issues here:
- Lines 74 - 76 should use `pending('Streaming not supported by user agent');` instead of an incorrect `done()` call.
- With #8768, it seems that the test can be enabled for other browsers (Chrome).

@yurydelendik Can the test be enabled for other browsers?

### PR DESCRIPTION
**Pull request by shikhar-scs**
As mentioned by @timvandermeij 3 days ago, I have made the required changes, also referring to the comments mentioned by @Snuffleupagus [here](https://github.com/mozilla/pdf.js/pull/9050/files/).

This hopefully closes #8851 .

### CODE REVIEW

File: test/unit/network_spec.js [hunk 1]
```diff
@@ -67,11 +67,15 @@ describe('network', function() {
...
-    var m = /Mozilla\/5.0.*?rv:(\d+).*? Gecko/.exec(userAgent);
-    if (!m || m[1] < 9) {
-      expect(true).toEqual(true);
-      done();
-      return;
+    var isFirefoxWithMozChunkedEncodingSupport =
+      /Mozilla\/5.0.*?rv:(?:\d+).*? Gecko/.test(userAgent);
+    let isFetchWithStreamSupport = (typeof Response !== 'undefined' &&
+      'body' in Response.prototype &&
+      typeof ReadableStream !== 'undefined');
```
Comment (Snuffleupagus) [inline test/unit/network_spec.js]: Nit: Please align these lines vertically, like so:
```javascript
let isFetchWithStreamSupport = (typeof Response !== 'undefined' &&
                                'body' in Response.prototype &&
                                typeof ReadableStream !== 'undefined');
```

File: test/unit/network_spec.js [hunk 2]
```diff
@@ -67,11 +67,15 @@ describe('network', function() {
...
+
+    if (!isFirefoxWithMozChunkedEncodingSupport &&
+      !isFetchWithStreamSupport) {
```
Comment (Snuffleupagus) [inline test/unit/network_spec.js]: Nit: Insufficient indentation on this line, please make sure that the two lines are aligned like below:
```javascript
if (!isFirefoxWithMozChunkedEncodingSupport &&
    !isFetchWithStreamSupport) {
```

Comment (shikhar-scs) [reply test/unit/network_spec.js]: Yeah Sure I'll do that rightway. Thanks for reviewing .

File: test/unit/network_spec.js [hunk 3]
```diff
@@ -67,11 +67,15 @@ describe('network', function() {
     // The test is valid for FF only: the XHR has support of the
     // 'moz-chunked-array' response type.
     // TODO enable for other browsers, e.g. when fetch/streams API is supported.
```
Comment (Snuffleupagus) [inline test/unit/network_spec.js]: As mentioned in https://github.com/mozilla/pdf.js/pull/9050/files/#r154503634, these comments are no longer necessary.

File: test/unit/network_spec.js [hunk 4]
```diff
@@ -67,11 +67,15 @@ describe('network', function() {
...
-    var m = /Mozilla\/5.0.*?rv:(\d+).*? Gecko/.exec(userAgent);
-    if (!m || m[1] < 9) {
-      expect(true).toEqual(true);
-      done();
-      return;
+    var isFirefoxWithMozChunkedEncodingSupport =
+      /Mozilla\/5.0.*?rv:(?:\d+).*? Gecko/.test(userAgent);
+    let isFetchWithStreamSupport = (typeof Response !== 'undefined' &&
+                                     'body' in Response.prototype &&
```
Comment (Snuffleupagus) [inline test/unit/network_spec.js]: There's one space too many on this line, please remove it.

Finally, you need to squash the commits when addressing review feedback; please see https://github.com/mozilla/pdf.js/wiki/Squashing-Commits.

File: test/unit/network_spec.js [hunk 5]
```diff
@@ -67,11 +67,15 @@ describe('network', function() {
     // The test is valid for FF only: the XHR has support of the
     // 'moz-chunked-array' response type.
     // TODO enable for other browsers, e.g. when fetch/streams API is supported.
-    var m = /Mozilla\/5.0.*?rv:(\d+).*? Gecko/.exec(userAgent);
-    if (!m || m[1] < 9) {
-      expect(true).toEqual(true);
-      done();
-      return;
+    var isFirefoxWithMozChunkedEncodingSupport =
```
Comment (timvandermeij) [inline test/unit/network_spec.js]: Change `var` to `let` here too.

**Comment by timvandermeij** [comment]
Please squash the commits into one commit after you've made these changes. Refer to https://github.com/mozilla/pdf.js/wiki/Squashing-Commits for how to do that easily.

**Comment by shikhar-scs** [comment]
@timvandermeij 
Sir I have changed that var to let and the commits have been squashed too.
@Snuffleupagus 
Sir, that extra white space was removed and unnecessary comments have been removed.

**Comment by shikhar-scs** [comment]
@Snuffleupagus Why is the unit tests failing in either of the OS?

TEST-UNEXPECTED-FAIL | read with streaming | Error: Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL.

TEST-UNEXPECTED-FAIL | read with streaming | Error: Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL.

**Comment by Snuffleupagus** [comment]
> Why is the unit tests failing in either of the OS?

I don't know offhand, but they seem to fail locally as well. Did you successfully run the unit-tests locally with this patch?

**Comment by shikhar-scs** [comment]
Initially while commiting I did and it was working perfectly. However after squashing i didnt perform the unit test. 

Also I tried commiting again in order to check if any difference has been made in any other file. However, 
that was not the case.

On branch enable-read-with-stream
nothing to commit, working directory clean


Running gulp unittest now renders this

gulp unittest
[15:01:32] Using gulpfile ~/Desktop/mozillaPdf/pdf.js/gulpfile.js
[15:01:32] Starting 'buildnumber'...

### Getting extension build number
[15:01:32] Starting 'locale'...

### Building localization files
Extension build number: 204
[15:01:32] Finished 'buildnumber' after 232 ms
[15:01:32] Starting 'components'...

### Creating generic components
[15:01:37] Version: webpack 3.10.0
            Asset    Size  Chunks             Chunk Names
    pdf_viewer.js  173 kB       0  [emitted]  main
pdf_viewer.js.map  272 kB       0  [emitted]  main
[15:01:37] Finished 'components' after 4.83 s
[15:01:37] Finished 'locale' after 5.2 s
[15:01:37] Starting 'generic'...

### Creating generic viewer

### Bundling files into pdf.js
[15:01:52] Version: webpack 3.10.0
        Asset    Size  Chunks                    Chunk Names
    viewer.js  364 kB       0  [emitted]  [big]  main
viewer.js.map  580 kB       0  [emitted]         main
[15:02:14] Version: webpack 3.10.0
     Asset    Size  Chunks                    Chunk Names
    pdf.js  601 kB       0  [emitted]  [big]  main
pdf.js.map  989 kB       0  [emitted]         main
[15:02:16] Version: webpack 3.10.0
            Asset     Size  Chunks                    Chunk Names
    pdf.worker.js   1.5 MB       0  [emitted]  [big]  main
pdf.worker.js.map  2.77 MB       0  [emitted]         main
[15:02:16] Finished 'generic' after 39 s
[15:02:16] Starting 'unittest'...

### Running unit tests
WARNING: File was not downloaded. See "pdfs/bug951051.pdf.error" file.
Unable to verify the checksum for the files that are used for testing.
Please re-download the files, or adjust the MD5 checksum in the manifest for the files listed above.

Server running at http://127.0.0.1:43871/
fs.js:1649
      binding.lstat(baseLong);
              ^

Error: ENOENT: no such file or directory, lstat '/path'
    at Object.realpathSync (fs.js:1649:15)
    at Function.WebBrowser.create (/home/shikhar/Desktop/mozillaPdf/pdf.js/test/webbrowser.js:250:17)
    at /home/shikhar/Desktop/mozillaPdf/pdf.js/test/test.js:643:30
    at Array.forEach (<anonymous>)
    at startBrowsers (/home/shikhar/Desktop/mozillaPdf/pdf.js/test/test.js:642:12)
    at startUnitTest (/home/shikhar/Desktop/mozillaPdf/pdf.js/test/test.js:564:3)
    at /home/shikhar/Desktop/mozillaPdf/pdf.js/test/test.js:734:7
    at /home/shikhar/Desktop/mozillaPdf/pdf.js/test/test.js:718:7
    at verifyNext (/home/shikhar/Desktop/mozillaPdf/pdf.js/test/downloadutils.js:151:7)
    at /home/shikhar/Desktop/mozillaPdf/pdf.js/test/downloadutils.js:178:7
[15:02:18] Finished 'unittest' after 1.48 s


Also I tried commiting again in order to check if any difference has been made in any other file. However, 
that was not the case.

On branch enable-read-with-stream
nothing to commit, working directory clean

**Comment by shikhar-scs** [comment]
An HTTP 504 error gets generated at "pdfs/bug951051.pdf.error" in the test folder.

**Comment by shikhar-scs** [comment]
WARNING: File was not downloaded. See "pdfs/bug951051.pdf.error" file.
Unable to verify the checksum for the files that are used for testing.
Please re-download the files, or adjust the MD5 checksum in the manifest for the files listed above. 

(sorry I dont know how to add a specific reply text)

I tried re downloading the file and ran the test again but to no avail. The same errors are repeating

**Comment by Snuffleupagus** [comment]
I'm not sure if errors quoted above are that relevant to the *unit* tests that we're concerned with here.

However, now that I look at the code in https://github.com/mozilla/pdf.js/blob/master/src/display/network.js, I'm not really sure how this unit-test would work in non-Firefox browsers, given that the `fetch` functionality is implemented in a *different* file; see https://github.com/mozilla/pdf.js/blob/master/src/display/fetch_stream.js.

Since the unit-tests in the `network_spec.js` file are only intended to test functionality in the `network.js` file, I now quite confused as to how this could actually work!?

@timvandermeij Any ideas here?

**Comment by shikhar-scs** [comment]
@Snuffleupagus I guess this new error that has occurred remained hidden initially because of the error which was reported in issue #8851 . Now that #8851 is corrected (hopefully) this error re-surfaced (which it originally should have had on its own).
I might be wrong though.

**Comment by timvandermeij** [comment]
The `fetch` functionality is indeed in a different file and ideally should also be unit tested in a different file for clarity. Looking at this particular test, it looks like it assumes the `fetch` functionality to work and rather focus on testing the `read` functionality of `PDFNetworkStream`, which is fine because there are different cases in which that should work (e.g., with and without streaming). I'm not sure though why the current patch wouldn't work.

@Rob--W Could you chime in here?

**Comment by shikhar-scs** [comment]
Any updates?

**Comment by Rob--W** [comment]
When I run the tests with this patch, it passes in Firefox but faiils in Chrome.

To help with debugging, run only this test, by starting a server with `gulp server` and then opening the following URL in Chrome: http://127.0.0.1:8888/test/unit/unit_test.html?spec=network%20read%20with%20streaming
Then you can edit the test and refresh the page to quickly see the effect of modifying `test/unit/network_spec.js`

I put a `console.log` after https://github.com/mozilla/pdf.js/blob/546cd2b67cade0ca5b083aff8346f4d50a40cd73/test/unit/network_spec.js#L42 and the log message is never being shown. This means that the promise returned by `getFullReader().read()` never resolves.

That promise is created here:
https://github.com/mozilla/pdf.js/blob/546cd2b67cade0ca5b083aff8346f4d50a40cd73/src/display/network.js#L459-L461

If I search in that file, the only location where this promise is fulfilled is at:

https://github.com/mozilla/pdf.js/blob/546cd2b67cade0ca5b083aff8346f4d50a40cd73/src/display/network.js#L387-L391

And `onProgressiveData` is only called in `_onDone` (= `args.onDone`) and in `args.onProgressiveData`:

https://github.com/mozilla/pdf.js/blob/546cd2b67cade0ca5b083aff8346f4d50a40cd73/src/display/network.js#L315-L322

These functions are used in `NetworkManager` (which contains logic that is conditional on `mozChunked` - note that `requestFull` in the next snippet is called by `)PDFNetworkStreamFullRequestReader`:
https://github.com/mozilla/pdf.js/blob/546cd2b67cade0ca5b083aff8346f4d50a40cd73/src/display/network.js#L86-L139

I stopped looking further, because I was expecting the test to use the `fetch` API in Chrome, and the above snippet clearly shows that `XMLHttpRequest` is used, which does not support streaming responses.

In Chrome, the following logic should have been called in the test, in order to test streaming.
https://github.com/mozilla/pdf.js/blob/546cd2b67cade0ca5b083aff8346f4d50a40cd73/src/display/fetch_stream.js#L44-L48

@yurydelendik wrote this unit test. Why is the test unconditionally using the `PDFNetworkStream` from `src/display/network.js`? (instead of feature-detecting the most appropriate `IPDFStream` implementation and testing that?)

**Comment by mukulmishra18** [comment]
> Why is the test unconditionally using the PDFNetworkStream from src/display/network.js? (instead of feature-detecting the most appropriate IPDFStream implementation and testing that?)

You are right. We are doing wrong here. We should detect for the support of `fetch/stream` and based on that set `PDFNetworkStream` just like we are doing here: https://github.com/mozilla/pdf.js/blob/546cd2b67cade0ca5b083aff8346f4d50a40cd73/src/pdf.js#L34-L44

If you look into `jasmine_boot.js` file: https://github.com/mozilla/pdf.js/blob/546cd2b67cade0ca5b083aff8346f4d50a40cd73/test/unit/jasmine-boot.js#L83-L89 we are setting `PDFNetworkStream` by feature-detection, but it is only applicable for `test/unit/api_spec.js`

**Comment by shikhar-scs** [comment]
@Snuffleupagus any updates?

**Comment by Snuffleupagus** [comment]
> @Snuffleupagus any updates?

Please read https://github.com/mozilla/pdf.js/pull/9268#issuecomment-351865602 and https://github.com/mozilla/pdf.js/pull/9268#issuecomment-351927462 above.

The gist of it though, is that the code that we're attempting to test here isn't implemented that way was initially assumed. Hence we need to figure out how to reconcile this *first*, probably in a separate issue/PR, before these unit-test can be enabled. So currently, it seems to me that this PR unfortunately needs to be put on hold for the time being.

**Comment by shikhar-scs** [comment]
@Snuffleupagus  Uhm Okay.

**Comment by Rob--W** [comment]
> Please read https://github.com/mozilla/pdf.js/pull/9268#issuecomment-351865602 and https://github.com/mozilla/pdf.js/pull/9268#issuecomment-351927462 above.
>
> The gist of it though, is that the code that we're attempting to test here isn't implemented that way was initially assumed. Hence we need to figure out how to reconcile this first, 

I would start with ensuring that `setPDFNetworkStreamClass` (as done in `jasmine-boot.js`) has run in all unit tests (or at least `network_spec.js`). This is fairly straightforward.

**Comment by timvandermeij** [comment]
Closing since the change is not in a mergable state at the moment and there has been no more follow-up. The issue will remain open and new pull requests to address the issue are welcome. For anyone willing to work on this, I suggest to read the comments above and make the necessary changes in small chunks (perhaps multiple pull requests). Thanks.

**Comment by shikhar-scs** [comment]
:disappointed:

**Pull request closed without merging**